assets/src/edit-story/components/canvas/carrousel/index.js

```diff
@@ -30,13 +31,47 @@ const Page = styled.a`
 `;
 
 function Canvas() {
-	const { state: { pages, currentPageIndex }, actions: { setCurrentPage } } = useStory();
-	const handleClickPage = ( page ) => () => setCurrentPage( { pageId: page.id } );
+	const { state: { pages, currentPageIndex }, actions: { setCurrentPage, arrangePage } } = useStory();
+	const getArrangeIndex = ( sourceIndex, dstIndex, position ) => {
+		// If the dropped element is before the dropzone index then we have to deduct
+		// that from the index to make up for the "lost" element in the row.
+		const indexAdjustment = sourceIndex < dstIndex ? -1 : 0;
+		if ( 'left' === position.x ) {
+			return dstIndex + indexAdjustment;
+		}
+		return dstIndex + 1 + indexAdjustment;
+	};
+	const handleClickPage = ( page ) => setCurrentPage( { pageId: page.id } );
 	return (
 		<List>
-			{ pages.map( ( page, index ) => (
-				<Page key={ index } onClick={ handleClickPage( page ) } isActive={ index === currentPageIndex } />
-			) ) }
+			{ pages.map( ( page, index ) => {
+				const onDrop = ( evt, position ) => {
+					const droppedEl = JSON.parse( evt.dataTransfer.getData( 'text' ) );
+					if ( ! droppedEl || 'page' !== droppedEl.type ) {
+						return;
+					}
+					const arrangedIndex = getArrangeIndex( droppedEl.index, index, position );
+					// Do nothing if the index didn't change.
+					if ( droppedEl.index !== arrangedIndex ) {
+						const pageId = pages[ droppedEl.index ].id;
+						arrangePage( { pageId, position: arrangedIndex } );
+						setCurrentPage( { pageId } );
+					}
+				};
+				// @todo Create a Draggable component for setting data and setting "draggable".
+				const onDragStart = ( evt ) => {
```

Comment:
Again, can't this be defined in a high scope.

Reply:
```
const onDragStart =  useCallback( ( index ) => ( evt ) => {
        const pageData = { type: 'page', 	index, };
	evt.dataTransfer.setData( 'text', JSON.stringify( pageData ) );
} ); 
```

```
<DropZone key={ index } onDrop={ onDrop( index ) } >
```

Reply:
Thanks! Done.